native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/sbom/SBOMGenerator.java

```diff
@@ -0,0 +1,220 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.graalvm.buildtools.maven.sbom;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.plugin.BuildPluginManager;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.logging.Logger;
+import org.eclipse.aether.RepositorySystem;
+import org.graalvm.buildtools.maven.NativeCompileNoForkMojo;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.twdata.maven.mojoexecutor.MojoExecutor.*;
+
+/**
+ * Generates an enhanced Software Bill of Materials (SBOM) for Native Image consumption and refinement.
+ * <p>
+ * Process overview:
+ * 1. Utilizes the cyclonedx-maven-plugin to create a baseline SBOM.
+ * 2. Augments the baseline SBOM components with additional metadata (see {@link AddedComponentFields}):
+ * * "packageNames": A list of all package names associated with each component.
+ * * "jarPath": Path to the component jar.
+ * * "prunable": Boolean indicating if the component can be pruned. We currently set this to false for
+ * any dependencies to the main component that are shaded.
+ * 3. Stores the enhanced SBOM at a known location.
+ * 4. Native Image then processes this SBOM during its static analysis:
+ * * Unreachable components are removed.
+ * * Unnecessary dependency relationships are pruned.
+ * <p>
+ * Creating the package-name-to-component mapping in the context of Native Image, without any build-system
+ * knowledge is difficult, which was the primary motivation for realizing this approach.
+ * <p>
+ * Benefits:
+ * * Great Baseline: Produces an industry-standard SBOM at minimum.
+ * * Enhanced Accuracy: Native Image static analysis refines the SBOM,
+ * potentially significantly improving its accuracy.
+ */
+final public class SBOMGenerator {
+    private final MavenProject mavenProject;
+    private final MavenSession mavenSession;
+    private final BuildPluginManager pluginManager;
+    private final RepositorySystem repositorySystem;
+    private final String mainClass;
+    private final Logger logger;
+
+    private static final String SBOM_NAME = "WIP_SBOM";
```

Comment:
I think this deserves a better name than "WIP". It's actually a proper SBOM based on build tooling information, which may be overapproximated, but useful on its own. How about "InitialSBOM" (to indicate that it'll be refined later on) or "BaselineSBOM"?

Reply:
I totally agree. I will update it to be called `base_sbom.json`.